File: 2/2.php
<?php

/**
 * Definition for a singly-linked list.
 */
class ListNode {
    public $val = 0;
    public $next = null;

    function __construct($val) {
        $this->val = $val;
    }
}

class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2) {
        $result = '';
        $tmp = &$result;
        $carry = 0;
        while (true) {
            if ($l1 == null && $l2 == null) {
                if ($carry) {
                    $tmp = new ListNode($carry);
                }
                break;
            }
            if ($l1 != null) {
                $value1 = $l1->val;
                $l1 = $l1->next;
            } else {
                $value1 = 0;
            }
            if ($l2 != null) {
                $value2 = $l2->val;
                $l2 = $l2->next;
            } else {
                $value2 = 0;
            }
            $sum = $value1 + $value2 + $carry;
            if ($sum >= 10) {
                $tmp = new ListNode($sum - 10);
                $carry = 1;
            } else {
                $tmp = new ListNode($sum);
                $carry = 0;
            }
            $tmp = &$tmp->next;
        }
        return $result;
    }
}

File: 2/2.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	l1 := &ListNode{1, nil}
	l2 := &ListNode{9, &ListNode{9, nil}}
	a := addTwoNumbers(l1, l2)
	fmt.Print(a)
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var new, next *ListNode
	carry := 0
	for l1 != nil || l2 != nil {
		tmp := 0
		tmp = tmp + carry
		carry = 0
		if l1 != nil {
			tmp = tmp + l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			tmp = tmp + l2.Val
			l2 = l2.Next
		}
		if tmp >= 10 {
			tmp = tmp - 10
			carry = 1
		}
		if next == nil {
			next = &ListNode{tmp, nil}
			new = next
		} else {
			next.Next = &ListNode{tmp, nil}
			next = next.Next
		}

	}
	if carry == 1 {
		next.Next = &ListNode{1, nil}
	}
	return new
}
